xtra-libp2p/src/endpoint.rs

```diff
@@ -526,26 +529,53 @@ impl Endpoint {
             "Type-system enforces that we only try to negotiate one protocol"
         );
 
-        let (protocol, stream) = self.open_substream(peer, protocols.clone()).await?;
+        let (control, _) = self.controls.get(&peer).ok_or(Error::NoConnection(peer))?;
 
-        debug_assert!(
-            protocol == protocols[0],
-            "If negotiation is successful, must have selected the only protocol we sent."
-        );
+        let fut = {
```

Comment:
Nice! Returning the future to the caller is a good one. And since we spin off tasks when looping over all connected peers this should solve this problem completely. 

Note: If we ever integrate a new network stack again or refactor on a bigger scale 😬😬 I would still try to thrive for a pattern where we have isolation per taker and not handle multiple takers in one actor (besides the low level connection handling in the endpoint).